lib/rspec/support/source/node.rb

```diff
@@ -0,0 +1,244 @@
+          :return0,
+          :sclass,
+          :stmts_add,
+          :stmts_new,
+          :string_add,
+          :string_concat,
+          :string_content,
+          :string_dvar,
+          :string_embexpr,
+          :string_literal,
+          :super,
+          :symbol,
+          :symbol_literal,
+          :symbols_add,
+          :symbols_new,
+          :top_const_field,
+          :top_const_ref,
+          :unary,
+          :undef,
+          :unless,
+          :unless_mod,
+          :until,
+          :until_mod,
+          :var_alias,
+          :var_field,
+          :var_ref,
+          :vcall,
+          :void_stmt,
+          :when,
+          :while,
+          :while_mod,
+          :word_add,
+          :word_new,
+          :words_add,
+          :words_new,
+          :xstring_add,
+          :xstring_literal,
+          :xstring_new,
+          :yield,
+          :yield0,
+          :zsuper
+        ]
+
+        attr_reader :sexp, :parent
+
+        def self.sexp?(array)
+          array.is_a?(Array) && array.first.is_a?(Symbol)
+        end
+
+        def initialize(ripper_sexp, parent=nil)
+          @sexp = ripper_sexp.freeze
+          @parent = parent
+        end
+
+        def type
+          sexp[0]
+        end
+
+        # Defines node type predicate methods.
+        # e.g. `node.fcall_type?`, which is equivalent to `node.type == :fcall`.
+        RIPPER_PARSER_EVENTS.each do |event|
+          define_method("#{event}_type?") do
+            type == event
+          end
+        end
+
+        # :@indent is not included in Ripper::PARSER_EVENTS.
+        def ident_type?
+          type == :@ident
+        end
+
+        def args
+          @args ||= raw_args.map do |raw_arg|
+            if Node.sexp?(raw_arg)
+              Node.new(raw_arg, self)
+            elsif Location.location?(raw_arg)
+              Location.new(*raw_arg)
+            elsif raw_arg.is_a?(Array)
+              ExpressionSequenceNode.new(raw_arg, self)
+            else
+              raw_arg
+            end
+          end.freeze
+        end
+
+        def children
+          @children ||= args.select { |arg| arg.is_a?(Node) }.freeze
+        end
+
+        def location
+          @location ||= args.find { |arg| arg.is_a?(Location) }
+        end
+
+        def each(&block)
+          return to_enum(__method__) unless block_given?
+
+          yield self
+
+          children.each do |child|
+            child.each(&block)
+          end
+        end
+
+        def each_ancestor
+          return to_enum(__method__) unless block_given?
+
+          current_node = self
+
+          while (current_node = current_node.parent)
+            yield current_node
+          end
+        end
+
+        def inspect
+          "#<#{self.class} #{type}>"
+        end
+
+      private
+
+        def raw_args
+          sexp[1..-1] || []
+        end
+      end
+
+      # @private
+      class ExpressionSequenceNode < Node
```

Comment:
I noticed rspec-core had `GroupNode` vs here you have `ExpressionSequenceNode`.  Can you explain why it needed to change?

BTW, a comment in the source explaining the need for this subclass would be helpful to our future selves, I think.